### PR TITLE
Scrapper for course data

### PR DESCRIPTION
**Pull request by PetterRein**
- add scraper for course data from "https://www.ntnu.no/studier/emner/" 

- add also if the exam from was digital or not
 
- updates faculties

- fix minor quality of life improvements

### CODE REVIEW

File: course_is_digital.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                if term is None:
+                    continue
+                
+                system: Optional[bs4.element.Tag] = dt.find(class_="exam-system")
+                exam_status: Optional[bs4.element.Tag] = dt.find(class_="exam-code")
+                
+                # skip continuation exam
+                #is_ordinary = exam_status.find("abbr").text.strip() == "ORD"
+                #if not is_ordinary:
+                #    continue
```
Comment (oleast) [inline course_is_digital.py]: Could we remove or handle this before merging?

File: course_is_digital.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                elif term_txt == "Høst":
+                    term_std = "Fall"
+                elif term_txt == "Sommer":
+                    term_std = "Summer"
+                else:
+                    continue
+                term_is_digital_dict[term_std] = system.text.strip() == "INSPERA"
+            result[year] = term_is_digital_dict
+            # sleep a little bit to avoid hammering the website
+            time.sleep(0.05)
```
Comment (oleast) [inline course_is_digital.py]: Shoudl we use the `sleep_time_mean_ms` here?

File: course_is_digital.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+def course_have_digital_exam_semester(course_code: str, year: str, semester_code: str):
+    semester_code = 'Spring' if semester_code == "V" else 'Fall' if semester_code == "H" else 'Summer'
+    years = []
+    years.append(year)
+    results = retrieve_exam_type_of_years(course_code, years)
+    if (results):
+        if (results[year].get(semester_code)):
+            return True
+        else:
+            return False
```
Comment (oleast) [inline course_is_digital.py]: Could this be reduced to just this?
``` python
return semester_code in results[year]
```

File: course_is_digital.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def course_have_digital_exam(course_code: str):
+    digitalExams = retrieve_exams_digital_course(course_code)
+    for i in range(2010,2020): # TODO: Probaly change the values here
+      exams_year = digitalExams[str(i)]
+      for key, value in exams_year.items(): 
+         if True == value: 
+             return True
+    return False
```
Comment (oleast) [inline course_is_digital.py]: Inconsistent casing

File: course_is_digital.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    #course_code = "TDT4127" # "TDT4100" TDT4117 IT3010 TDT4127 (itgk)
+    #years = retrieve_exam_years(course_code)
+    #print(retrieve_exam_type_of_years(course_code, years))
+    #print(course_have_digital_exam("IT3010"))
+    print(course_have_digital_exam_semester("TDT4100", 2018, "H"))
+
+
+
+if __name__ == '__main__':
+    _test_functions()
```
Comment (oleast) [inline course_is_digital.py]: Could we remove this before merging?

File: grades.py [hunk 6]
```diff
@@ -44,11 +46,13 @@ def login(username, password):
 
 
 def create_course(code, faculty):
-    base_url = "http://www.ime.ntnu.no/api/course/"
-    resp = requests.get(url=base_url + code)
+    #base_url = "http://www.ime.ntnu.no/api/course/"
+    #resp = requests.get(url=base_url + code)
```
Comment (oleast) [inline grades.py]: Should probably remove these old API thingies?

File: grades.py [hunk 7]
```diff
@@ -44,11 +46,13 @@ def login(username, password):
...
+    resp = getCourseData(code, faculty)
     if not resp:
         return None
-    data = json.loads(resp.text)
+    #data = json.loads(resp.text)
```
Comment (oleast) [inline grades.py]: Remove?

File: course_is_digital.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                else:
+                    continue
+                term_is_digital_dict[term_std] = system.text.strip() == "INSPERA"
+            result[year] = term_is_digital_dict
+            # sleep a little bit to avoid hammering the website
+            time.sleep(0.05)
+        return result
+
+
+def course_have_digital_exam_semester(course_code: str, year: str, semester_code: str):
```
Comment (oleast) [inline course_is_digital.py]: Should probably be `course_has_digital_exam`?

File: grades/models.py [hunk 9]
```diff
@@ -40,16 +40,20 @@ def course_level(self):
     def __unicode__(self):
         return self.code
     
-    @permalink
+    def __str__(self):
+        return self.code
+
+    # FIX Permalink depercated
```
Comment (oleast) [inline grades/models.py:50]: Do you know what was the actual use for this?

File: scrapeCourse.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    except ValueError:
+      print("Not valid number, number is: ", facts_about_course[2].split("\n")[2])
+    norwegianNameRaw = soup_no.title.get_text().split("-")[1]
+    norwegianName = norwegianNameRaw[1:len(norwegianNameRaw)-1]
+    englishNameRaw = soup_eng.title.get_text().split("-")[1]
+    englishName = englishNameRaw[1:len(englishNameRaw)-1]
+    courseLevelText = facts_about_course[3].split("\n")[0][1:len(facts_about_course[3].split("\n")[0])]
+    study_level = courseLevel(courseLevelText, code)
+    last_year_taught = 0
+    taught_from = 2008
```
Comment (oleast) [inline scrapeCourse.py]: Why is this set to 2008?

File: grades/models.py [hunk 11]
```diff
@@ -40,16 +40,20 @@ def course_level(self):
...
+        return reverse('course', kwargs={'course_code': self.code})
 
 
 class Grade(models.Model):
-    course = models.ForeignKey(Course)
+    course = models.ForeignKey(Course, on_delete=models.CASCADE)
     semester_code = models.CharField("Semester", max_length=10)
     
     average_grade = models.FloatField()
+    digitalExam = models.BooleanField(default=False)
```
Comment (oleast) [inline grades/models.py]: Should be `digital_exam` to keep up with Python naming practices.

File: scrapeCourse.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
+import os
+import json
+import getpass
+import django
+import json
+import re
+from bs4 import BeautifulSoup
+import requests
+
+def getCourseData(code, faculty):
```
Comment (oleast) [inline scrapeCourse.py:28]: This file has inconsistent indentation with the rest of the project

**Review by oleast — COMMENTED** [review]
Awesome work!! :heart_eyes: 

I added a few comments, but it is mostly nitpicking :smile: 

I have not been able to test it for myself yet, but I'll have a look at it when I can :smiley:

File: course_is_digital.py (hunk 1 above)
Comment (PetterRein) [reply course_is_digital.py]: Yes

File: course_is_digital.py (hunk 2 above)
Comment (PetterRein) [reply course_is_digital.py]: Probably, was testing with time.sleep

File: course_is_digital.py (hunk 3 above)
Comment (PetterRein) [reply course_is_digital.py]: Jupp

File: course_is_digital.py (hunk 4 above)
Comment (PetterRein) [reply course_is_digital.py]: casing? line 118?

File: course_is_digital.py (hunk 5 above)
Comment (PetterRein) [reply course_is_digital.py]: Yup

File: scrapeCourse.py (hunk 10 above)
Comment (PetterRein) [reply scrapeCourse.py]: Every course on "emner" starts from 2008, but I know some courses started before that but we cannot get the data. taught_from is not used anywhere so maybe we could remove it

File: grades/models.py (hunk 11 above)
Comment (PetterRein) [reply grades/models.py]: Yupp

File: scrapeCourse.py (hunk 12 above)
Comment (PetterRein) [reply scrapeCourse.py:28]: Will fix

File: grades/models.py (hunk 9 above)
Comment (PetterRein) [reply grades/models.py:50]: Not really, only that it was used for reverse the URL of courses

File: course_is_digital.py (hunk 8 above)
Comment (PetterRein) [reply course_is_digital.py]: Dont think so

**Review by oleast — APPROVED** [review]
Looks good to me :heart_eyes:

**Pull request merged**